### PR TITLE
feat: moved trace logic to a server.onLoad hook

### PR DESCRIPTION
**Pull request by ResuBaka**
### Related issues

closes #

### Which environment this relates to
Check your case. In case of any doubts please read about [Release Cycle](https://docs.vuestorefront.io/guide/basics/release-cycle.html)

- [ ] Test version (https://test.storefrontcloud.io) - this is a new feature or improvement for Vue Storefront. I've created branch from `develop` branch and want to merge it back to `develop`
- [ ] RC version (https://next.storefrontcloud.io) - this is a stabilisation fix for Release Candidate of Vue Storefront. I've created branch from `release` branch and want to merge it back to `release`
- [ ] Stable version (https://demo.storefrontcloud.io) - this is an important fix for current stable version. I've created branch from `hotfix` or `master` branch and want to merge it back to `hotfix`

### Upgrade Notes and Changelog

- [x] No upgrade steps required (100% backward compatibility and no breaking changes)
- [ ] I've updated the [Upgrade notes](https://github.com/DivanteLtd/vue-storefront/blob/develop/docs/guide/upgrade-notes/README.md) and [Changelog](https://github.com/DivanteLtd/vue-storefront/blob/develop/CHANGELOG.md) on how to port existing VS sites with this new feature

**IMPORTANT NOTICE** - Remember to update `CHANGELOG.md` with description of your change

### Contribution and currently important rules acceptance

- [ ] I read and followed [contribution rules](https://github.com/DivanteLtd/vue-storefront/blob/master/CONTRIBUTING.md)

### CODE REVIEW

**Comment by ResuBaka** [comment]
hey @filrak you mean something like that? 

When yes I would polish it a bit more and maybe move the whole serverModules logic to a hook to so you can just have different exports on like modules two for the server then. One that is loaded pretty early like now the tracing stuff needs to be and the second one for added routes or middlewares to the server.

**Comment by ResuBaka** [comment]
@pkarw 

I have moved the whole extend logic to a hook which takes the `app`, `config.server` and `isProd` as an object as a parameter. 
Then I have changed the first hook to a be named trace hook so you know that is should be used to add your tracing library.

**Comment by pkarw** [comment]
@ResuBaka would be awesome to add some additional hooks by this chance:
- afterOutputRendered - so the users can adjust req/res/context doing some filtering etc - by having this we could have a Munich plugin not in the core as we currently have but asa separate module 
- beforeOutputRendered - passing the request context that can be modified / prepared + res / req so the users can adjust the headers or the context  - we can even do something like if the hook will return any continent / set any content or pass some flag or do req.end we can skip running ssr - I mean the way to use this hook to implement custom caching algorithm
- beforeCacheInvalidated

**Comment by ResuBaka** [comment]
I have added the new hooks but that is only a basic implementation as I am not 100% sure how to add each of them in the best way. 

So it would be great when I get there some feedback on what they should be able to do.

**Comment by filrak** [comment]
@pkarw at this point before adding those hooks n89mI want you to decide if we want to be server-comaptible with nuxt in terms of hooks. The ones you proposed cannot be translated to other frameworks therefore will break **potentially** all server side modules. 

Are we okay with that? I think if you see urgent need for them now we can be ok with that because we will drop support for those modules but have nuxt ones in return.

If you don't see urgent need for this hooks I would advice to wait with adding  (any) new APIs until they needed.

**Comment by filrak** [comment]
@ResuBaka I'll wait with a. review till @pkarw will answer because it takes into account function signatures as well

File: core/scripts/server.ts [hunk 1]
```diff
@@ -71,6 +72,9 @@ function invalidateCache (req, res) {
         tags = req.query.tag.split(',')
       }
       const subPromises = []
+
+      serverHooksExecutors.beforeCacheInvalidated()
```
Comment (pkarw) [inline core/scripts/server.ts]: let's pass the all parameters in here nad it's fine

File: core/scripts/server.ts [hunk 2]
```diff
@@ -186,6 +202,15 @@ app.get('*', (req, res, next) => {
           tagsArray
         ).catch(errorHandler)
       }
+
+      const hookResponse2 = serverHooksExecutors.afterOutputRendered({
```
Comment (pkarw) [inline core/scripts/server.ts]: Rename to `afterOutputRenderedResponse`

File: core/scripts/server.ts [hunk 3]
```diff
@@ -178,6 +185,15 @@ app.get('*', (req, res, next) => {
         res.setHeader('X-VS-Cache-Tags', cacheTags)
         console.log(`cache tags for the request: ${cacheTags}`)
       }
+
+      const hookResponse = serverHooksExecutors.beforeOutputRendered({
```
Comment (pkarw) [inline core/scripts/server.ts]: Rename to `beforeOutputRenderedResponse`

File: core/scripts/server.ts [hunk 4]
```diff
@@ -186,6 +202,15 @@ app.get('*', (req, res, next) => {
...
+
+      const hookResponse2 = serverHooksExecutors.afterOutputRendered({
+        req,
+        res,
+        context,
+        output,
+        isProd
+      })
+
       res.end(output)
```
Comment (pkarw) [inline core/scripts/server.ts]: if `afterOutputRenderedResponse` !== null it should be then used instead of `output`

Comment (ResuBaka) [reply core/scripts/server.ts]: I have used this logic here: 

```typescript
      if (typeof afterOutputRenderedResponse.output === 'string') {
        res.end(afterOutputRenderedResponse.output)
      } else if (typeof afterOutputRenderedResponse === 'string') {
        res.end(afterOutputRenderedResponse)
      } else {
        res.end(output)
      }
```

As this should be more what we want I think.

**Comment by ResuBaka** [comment]
This PR should not be merged to fast. 

As we should exactly know what each hook should do. As the MutatorHook I think needs a bit more thinking if there are right or not

**Comment by andrzejewsky** [comment]
@ResuBaka is it ready to be merged in? Are you still working on this?

**Comment by ResuBaka** [comment]
@andrzejewsky it should work. I have not tested it, I only have written it after the logic of how each hook should work like from there internal logic.  

So I would say that before we merge it there should be done a bit of QA of each hook so it we are 100% sure that it all works like every buddy would expect this works.

**Pull request merged**